x-pack/plugins/security/server/routes/authorization/roles/put.ts

```diff
@@ -0,0 +1,252 @@
+import { GLOBAL_RESOURCE } from '../../../../common/constants';
+import { wrapError } from '../../../errors';
+import { PrivilegeSerializer } from '../../../authorization/privilege_serializer';
+import { ResourceSerializer } from '../../../authorization/resource_serializer';
+import { ElasticsearchRole } from '.';
+
+export function definePutRolesRoutes({ router, authz, clusterClient }: RouteDefinitionParams) {
+  const transformPrivilegesToElasticsearchPrivileges = (
+    privileges: TypeOf<typeof payloadSchema>['kibana'] = []
+  ) => {
+    return privileges.map(({ base, feature, spaces }) => {
+      if (spaces.length === 1 && spaces[0] === GLOBAL_RESOURCE) {
+        return {
+          privileges: [
+            ...(base
+              ? base.map(privilege => PrivilegeSerializer.serializeGlobalBasePrivilege(privilege))
+              : []),
+            ...(feature
+              ? Object.entries(feature)
+                  .map(([featureName, featurePrivileges]) =>
+                    featurePrivileges.map(privilege =>
+                      PrivilegeSerializer.serializeFeaturePrivilege(featureName, privilege)
+                    )
+                  )
+                  .flat()
+              : []),
+          ],
+          application: authz.getApplicationName(),
+          resources: [GLOBAL_RESOURCE],
+        };
+      }
+
+      return {
+        privileges: [
+          ...(base
+            ? base.map(privilege => PrivilegeSerializer.serializeSpaceBasePrivilege(privilege))
+            : []),
+          ...(feature
+            ? Object.entries(feature)
+                .map(([featureName, featurePrivileges]) =>
+                  featurePrivileges.map(privilege =>
+                    PrivilegeSerializer.serializeFeaturePrivilege(featureName, privilege)
+                  )
+                )
+                .flat()
+            : []),
+        ],
+        application: authz.getApplicationName(),
+        resources: (spaces as string[]).map(resource =>
+          ResourceSerializer.serializeSpaceResource(resource)
+        ),
+      };
+    });
+  };
+
+  const transformRolesToElasticsearchRoles = (
+    rolePayload: TypeOf<typeof payloadSchema>,
+    existingApplications: ElasticsearchRole['applications'] = []
+  ) => {
+    const {
+      elasticsearch = { cluster: undefined, indices: undefined, run_as: undefined },
+      kibana = [],
+    } = rolePayload;
+    const otherApplications = existingApplications.filter(
+      roleApplication => roleApplication.application !== authz.getApplicationName()
+    );
+
+    return {
+      metadata: rolePayload.metadata,
+      cluster: elasticsearch.cluster || [],
+      indices: elasticsearch.indices || [],
+      run_as: elasticsearch.run_as || [],
+      applications: [...transformPrivilegesToElasticsearchPrivileges(kibana), ...otherApplications],
+    } as Omit<ElasticsearchRole, 'name'>;
+  };
+
+  const FEATURE_NAME_VALUE_REGEX = /^[a-zA-Z0-9_-]+$/;
```

Comment:
__note:__ with such a complex schema I'm heavily relying on various tests we have :)